be/src/fs/s3/poco_common.cpp

```diff
@@ -97,4 +81,98 @@ bool checkRequestCanReturn2xxAndErrorInBody(Aws::Http::HttpRequest& request) {
+    HTTPSessionPtr session;
+
+    if (https) {
+        session = std::make_shared<Poco::Net::HTTPSClientSession>(host, port);
+    } else {
+        session = std::make_shared<Poco::Net::HTTPClientSession>(host, port);
+    }
+
+    // doesn't work properly without patch
+    session->setKeepAlive(keep_alive);
+    return session;
+}
+
+class EndpointHTTPSessionPool : public PoolBase<Poco::Net::HTTPClientSession> {
+public:
+    using Base = PoolBase<Poco::Net::HTTPClientSession>;
+    EndpointHTTPSessionPool(const std::string& host, uint16_t port, bool is_https)
+            : Base(ENDPOINT_POOL_SIZE), _host(host), _port(port), _is_https(is_https) {}
+
+private:
+    ObjectPtr allocObject() override {
+        SCOPED_THREAD_LOCAL_MEM_TRACKER_SETTER(GlobalEnv::GetInstance()->poco_connection_pool_mem_tracker());
+        auto session = makeHTTPSessionImpl(_host, _port, _is_https, true);
+        return session;
+    }
+
+    static constexpr size_t ENDPOINT_POOL_SIZE = 50;
+
+    const std::string _host;
+    const uint16_t _port;
+    const bool _is_https;
+};
+
+class HTTPSessionPools {
+private:
+    struct Key {
+        std::string host;
+        uint16_t port;
+        bool is_https;
+
+        bool operator==(const Key& rhs) const {
+            return std::tie(host, port, is_https) == std::tie(rhs.host, rhs.port, rhs.is_https);
+        }
+    };
+
+    struct Hasher {
+        uint32_t operator()(const Key& k) const {
+            return std::hash<std::string>()(k.host) | std::hash<uint16_t>()(k.port) | std::hash<bool>()(k.is_https);
+        }
+    };
+
+public:
+    using EndpointPoolPtr = std::shared_ptr<EndpointHTTPSessionPool>;
+    static HTTPSessionPools& instance() {
+        static HTTPSessionPools instance;
+        return instance;
+    }
+
+    PooledHTTPSessionPtr getSession(const Poco::URI& uri, const ConnectionTimeouts& timeouts, bool resolve_host) {
+        const std::string& host = uri.getHost();
+        uint16_t port = uri.getPort();
+        bool is_https = isHTTPS(uri);
+
+        const Key key(host, port, is_https);
+
+        EndpointPoolPtr pool = nullptr;
+        {
+            std::lock_guard lock(_mutex);
+            auto item = _endpoint_pools.find(key);
+            if (item == _endpoint_pools.end()) {
+                std::tie(item, std::ignore) =
+                        _endpoint_pools.emplace(key, std::make_shared<EndpointHTTPSessionPool>(host, port, is_https));
+            }
+            pool = item->second;
+        }
+
+        auto session = pool->get(timeouts.connection_timeout.totalMicroseconds());
+        session->attachSessionData({});
+
+        return session;
+    }
+
+private:
+    HTTPSessionPools() = default;
+
+    std::mutex _mutex;
+    std::unordered_map<Key, EndpointPoolPtr, Hasher> _endpoint_pools;
+};
+
+PooledHTTPSessionPtr makeHTTPSession(const Poco::URI& uri, const ConnectionTimeouts& timeouts, bool resolve_host) {
```

Comment:
need to provide an interface to properly shutdown the singleton, for a clean BE shutdown.

Reply:
will do it